# **Student Applications**

This notebook focuses on processing student application data. The goal is to clean and prepare the dataset for analysis — specifically to identify students who are compatible with the Accelerator Data Science Bootcamp.

## Import dependencies:

In [1]:
import pandas as pd

## Load the data:

In [2]:
applications_df = pd.read_csv("../data/skills_test_student_applications.csv")

## Data exploration functions:

In [3]:
def column_missing_values(df, column_name):
    if df[column_name].isnull().sum() > 0:
        if df[column_name].isnull().sum() == df.shape[0]:
            return f"'{column_name}' is empty."
        else:
            return column_name

In [4]:
def get_total_columns_missing_values(df):
    columns_with_missing_values = []

    for column_name in df.columns:
        missing_values_column = column_missing_values(df, column_name)
        if missing_values_column:
            columns_with_missing_values.append(missing_values_column)

    return len(columns_with_missing_values), columns_with_missing_values

In [5]:
def column_type(df, column_name):
    return df[column_name].dtype

## Assertion functions:

In [6]:
def is_column_values_unique(df, column_name):
    assert df[
        column_name
    ].is_unique, f"Expected True, meaning all values are unique, but got '{df[column_name].is_unique}', which means all values are not unique."

## Explore the data:

In [7]:
pd.set_option("display.max_columns", None)

In [8]:
applications_df.head()

Application ID  Learner ID Gender           Race    Nationality  \
0           41581         572    Man  African/Black  South African   
1           73815         857  Woman  African/Black  South African   
2           46124         908    Man  African/Black  South African   
3          254193         928    Man  African/Black  South African   
4           76156        1132  Woman  African/Black  South African   

  Country of residence         Registration date          Application date  \
0         South Africa  2024-05-11T13:27:10.490Z  2025-03-22T12:03:12.669Z   
1         South Africa  2024-05-12T18:55:56.670Z  2025-07-21T07:32:00.689Z   
2         South Africa  2024-05-13T03:47:02.685Z  2025-04-15T22:45:48.117Z   
3         South Africa  2024-05-13T06:12:08.009Z  2025-09-19T13:47:28.466Z   
4         South Africa  2024-05-13T14:32:21.448Z  2025-08-03T08:55:31.590Z   

   Total score  Literacy score  Logic score  Sequence score  \
0           48               8            9               7   
1           51               9            8               9   
2           43               9            7               6   
3           48               9            9               7   
4           57               9           10               9   

   Problem solving score  Numeracy score  Statistics score  \
0                      7               7                10   
1                      9               8                 8   
2                      3               9                 9   
3                      9               9                 5   
4                     10              10                 9   

   Has Taken Aptitude Test Aptitude test date attempted Has Coding experience  \
0                      NaN     2025-03-30T19:03:46.734Z                 False   
1                      NaN     2024-05-21T21:04:48.489Z                  True   
2                      NaN     2025-04-16T11:19:07.408Z                 False   
3                      NaN     2024-05-13T07:52:28.779Z                  True   
4                      NaN     2025-08-03T09:12:32.861Z                  True   

    Status                     Programming language experince  \
0  Applied                                             [null]   
1  Applied           ["C++","HTML/CSS","JavaScript","Python"]   
2  Applied                                             [null]   
3  Applied  ["HTML/CSS","JavaScript","Other (please specif...   
4  Applied                       ["HTML/CSS","Java","Python"]   

  Code Challenge plagirism  Code challenge time taken(min)  \
0                      NaN                             NaN   
1                      NaN                             NaN   
2                      NaN                             NaN   
3                      NaN                             NaN   
4                      NaN                             NaN   

   Coding challenge completed  Coding challenge score  \
0                         NaN                     NaN   
1                         NaN                     NaN   
2                         NaN                     NaN   
3                         NaN                     NaN   
4                         NaN                     NaN   

   Code challenge multiple choice score  Code challenge final score  \
0                                   NaN                         NaN   
1                                   NaN                         NaN   
2                                   NaN                         NaN   
3                                   NaN                         NaN   
4                                   NaN                         NaN   

   Code challenge cheat tab leaving (no.) Code challenge cheat plagiarism  \
0                                     NaN                             NaN   
1                                     NaN                             NaN   
2                                     NaN                             NaN   
3                               

In [9]:
applications_df.shape

(599, 81)

Initial number of `applications_df` rows is `599`, and `applications_df` columns is `81`.

In [10]:
is_column_values_unique(applications_df, "Application ID")

In [11]:
initial_rows_number = applications_df.shape[0]
initial_columns_number = applications_df.shape[1]

In [12]:
applications_df.duplicated(subset="Application ID").sum()

np.int64(0)

There are no duplicate rows.

In [13]:
total_columns, columns_with_missing_values = get_total_columns_missing_values(
    applications_df
)

In [14]:
empty_columns = []

for column_name in columns_with_missing_values:
    if "is empty." in column_name:
        empty_columns.append(column_name)

f"There are {len(empty_columns)} empty columns:{empty_columns}"

'There are 13 empty columns:["\'Has Taken Aptitude Test\' is empty.", "\'has_completed_application\' is empty.", "\'application_accepted\' is empty.", "\'is_south_african_citizen\' is empty.", "\'learner_referral\' is empty.", "\'other_specify_internet_access_type\' is empty.", "\'other_programming_language_specify\' is empty.", "\'has_criminal_record\' is empty.", "\'nearest_city\' is empty.", "\'is_currently_employed\' is empty.", "\'other_race_specify\' is empty.", "\'asresidential_area_type\' is empty.", "\'career_stream.1\' is empty."]'

Out of 48 columns that have missing values, 13 are empty.

## Manipulate data: 

In [15]:
applications_df = applications_df.dropna(axis=1, how="all")

In [16]:
applications_df.shape

(599, 68)

In [17]:
total_columns, columns_with_missing_values = get_total_columns_missing_values(
    applications_df
)
empty_columns = []

for column_name in columns_with_missing_values:
    if "is empty." in column_name:
        empty_columns.append(column_name)

f"There are {len(empty_columns)} empty columns:{empty_columns}"

'There are 0 empty columns:[]'

In [18]:
applications_df.head()

Application ID  Learner ID Gender           Race    Nationality  \
0           41581         572    Man  African/Black  South African   
1           73815         857  Woman  African/Black  South African   
2           46124         908    Man  African/Black  South African   
3          254193         928    Man  African/Black  South African   
4           76156        1132  Woman  African/Black  South African   

  Country of residence         Registration date          Application date  \
0         South Africa  2024-05-11T13:27:10.490Z  2025-03-22T12:03:12.669Z   
1         South Africa  2024-05-12T18:55:56.670Z  2025-07-21T07:32:00.689Z   
2         South Africa  2024-05-13T03:47:02.685Z  2025-04-15T22:45:48.117Z   
3         South Africa  2024-05-13T06:12:08.009Z  2025-09-19T13:47:28.466Z   
4         South Africa  2024-05-13T14:32:21.448Z  2025-08-03T08:55:31.590Z   

   Total score  Literacy score  Logic score  Sequence score  \
0           48               8            9               7   
1           51               9            8               9   
2           43               9            7               6   
3           48               9            9               7   
4           57               9           10               9   

   Problem solving score  Numeracy score  Statistics score  \
0                      7               7                10   
1                      9               8                 8   
2                      3               9                 9   
3                      9               9                 5   
4                     10              10                 9   

  Aptitude test date attempted Has Coding experience   Status  \
0     2025-03-30T19:03:46.734Z                 False  Applied   
1     2024-05-21T21:04:48.489Z                  True  Applied   
2     2025-04-16T11:19:07.408Z                 False  Applied   
3     2024-05-13T07:52:28.779Z                  True  Applied   
4     2025-08-03T09:12:32.861Z                  True  Applied   

                      Programming language experince Code Challenge plagirism  \
0                                             [null]                      NaN   
1           ["C++","HTML/CSS","JavaScript","Python"]                      NaN   
2                                             [null]                      NaN   
3  ["HTML/CSS","JavaScript","Other (please specif...                      NaN   
4                       ["HTML/CSS","Java","Python"]                      NaN   

   Code challenge time taken(min)  Coding challenge completed  \
0                             NaN                         NaN   
1                             NaN                         NaN   
2                             NaN                         NaN   
3                             NaN                         NaN   
4                             NaN                         NaN   

   Coding challenge score  Code challenge multiple choice score  \
0                     NaN                                   NaN   
1                     NaN                                   NaN   
2                     NaN                                   NaN   
3                     NaN                                   NaN   
4                     NaN                                   NaN   

   Code challenge final score  Code challenge cheat tab leaving (no.)  \
0                         NaN                                     NaN   
1                         NaN                                     NaN   
2                         NaN                                     NaN   
3                         NaN                                     NaN   
4                         NaN                                     NaN   

  Code challenge cheat plagiarism Code challenge cheat pasted code  \
0                             NaN                              NaN   
1                             NaN                              NaN   
2                             NaN       

In [19]:
lower_columns = []

for column_name in applications_df.columns:
    if column_name == "Total score":
        column_name = "Aptitude Total Score"
    elif column_name == "Programming language experince":
        column_name = "Programming language experience"

    lower_columns.append(column_name.lower().replace(" ", "_"))

applications_df.columns = lower_columns
applications_df.head(0)

Empty DataFrame
Columns: [application_id, learner_id, gender, race, nationality, country_of_residence, registration_date, application_date, aptitude_total_score, literacy_score, logic_score, sequence_score, problem_solving_score, numeracy_score, statistics_score, aptitude_test_date_attempted, has_coding_experience, status, programming_language_experience, code_challenge_plagirism, code_challenge_time_taken(min), coding_challenge_completed, coding_challenge_score, code_challenge_multiple_choice_score, code_challenge_final_score, code_challenge_cheat_tab_leaving_(no.), code_challenge_cheat_plagiarism, code_challenge_cheat_pasted_code, code_challenge_cheat_suspicious_activity, code_challenge_cheat_ai_usage, coding_experience_description, other_programming_language_experince_specify, learner_holds_coding_certs, learner_coding_certs_list, learner_program, learner_career_streams, pathways, how_far_away_from_city, internet_access_type, residential_area_type, internet_data_access_type, other_specify_residential_area_type, learner_comfortable_disclosing_income, income_currency, learner_monthly_income, has_yoma_platform, computer_access_type, home_language, has_disability_or_differently_abled, is_refugee, preferred_language_of_instruction, employment_type, english_proficiency, second_language, highest_education_level, field_of_current_education, is_currently_completing_education, ever_regisitered_for_yes4youth, is_umuzi_alumnus, current_institution_study, currently_studying, institution_highest_education, qualification_highest_education, career_stream, age_formula, age_range, age_range_order, application_status]
Index: []

In [20]:
applications_df.columns.duplicated().sum()

np.int64(0)

In [21]:
mandatory_columns = (
    "application_id",
    "learner_id",
    "gender",
    "race",
    "nationality",
    "country_of_residence",
    "registration_date",
    "application_date",
    "aptitude_total_score",
    "qualification_highest_education",
    "currently_studying",
    "field_of_current_education",
    "highest_education_level",
    "programming_language_experience",
    "other_programming_language_experince_specify",
    "has_coding_experience",
    "numeracy_score",
    "logic_score",
    "problem_solving_score",
    "sequence_score",
    "literacy_score",
    "statistics_score",
    "learner_holds_coding_certs",
    "learner_coding_certs_list",
    "learner_program",
    "learner_career_streams",
    "career_stream",
    "age_formula",
    "age_range",
    "age_range_order",
    "employment_type",
    "english_proficiency",
    "computer_access_type",
)

In [22]:
non_mandatory = []

for column_name in applications_df.columns.to_list():
    if str(column_name) not in mandatory_columns:
        non_mandatory.append(column_name)

non_mandatory

['aptitude_test_date_attempted',
 'status',
 'code_challenge_plagirism',
 'code_challenge_time_taken(min)',
 'coding_challenge_completed',
 'coding_challenge_score',
 'code_challenge_multiple_choice_score',
 'code_challenge_final_score',
 'code_challenge_cheat_tab_leaving_(no.)',
 'code_challenge_cheat_plagiarism',
 'code_challenge_cheat_pasted_code',
 'code_challenge_cheat_suspicious_activity',
 'code_challenge_cheat_ai_usage',
 'coding_experience_description',
 'pathways',
 'how_far_away_from_city',
 'internet_access_type',
 'residential_area_type',
 'internet_data_access_type',
 'other_specify_residential_area_type',
 'learner_comfortable_disclosing_income',
 'income_currency',
 'learner_monthly_income',
 'has_yoma_platform',
 'home_language',
 'has_disability_or_differently_abled',
 'is_refugee',
 'preferred_language_of_instruction',
 'second_language',
 'is_currently_completing_education',
 'ever_regisitered_for_yes4youth',
 'is_umuzi_alumnus',
 'current_institution_study',
 'inst

In [23]:
applications_df.head()

application_id  learner_id gender           race    nationality  \
0           41581         572    Man  African/Black  South African   
1           73815         857  Woman  African/Black  South African   
2           46124         908    Man  African/Black  South African   
3          254193         928    Man  African/Black  South African   
4           76156        1132  Woman  African/Black  South African   

  country_of_residence         registration_date          application_date  \
0         South Africa  2024-05-11T13:27:10.490Z  2025-03-22T12:03:12.669Z   
1         South Africa  2024-05-12T18:55:56.670Z  2025-07-21T07:32:00.689Z   
2         South Africa  2024-05-13T03:47:02.685Z  2025-04-15T22:45:48.117Z   
3         South Africa  2024-05-13T06:12:08.009Z  2025-09-19T13:47:28.466Z   
4         South Africa  2024-05-13T14:32:21.448Z  2025-08-03T08:55:31.590Z   

   aptitude_total_score  literacy_score  logic_score  sequence_score  \
0                    48               8            9               7   
1                    51               9            8               9   
2                    43               9            7               6   
3                    48               9            9               7   
4                    57               9           10               9   

   problem_solving_score  numeracy_score  statistics_score  \
0                      7               7                10   
1                      9               8                 8   
2                      3               9                 9   
3                      9               9                 5   
4                     10              10                 9   

  aptitude_test_date_attempted has_coding_experience   status  \
0     2025-03-30T19:03:46.734Z                 False  Applied   
1     2024-05-21T21:04:48.489Z                  True  Applied   
2     2025-04-16T11:19:07.408Z                 False  Applied   
3     2024-05-13T07:52:28.779Z                  True  Applied   
4     2025-08-03T09:12:32.861Z                  True  Applied   

                     programming_language_experience code_challenge_plagirism  \
0                                             [null]                      NaN   
1           ["C++","HTML/CSS","JavaScript","Python"]                      NaN   
2                                             [null]                      NaN   
3  ["HTML/CSS","JavaScript","Other (please specif...                      NaN   
4                       ["HTML/CSS","Java","Python"]                      NaN   

   code_challenge_time_taken(min)  coding_challenge_completed  \
0                             NaN                         NaN   
1                             NaN                         NaN   
2                             NaN                         NaN   
3                             NaN                         NaN   
4                             NaN                         NaN   

   coding_challenge_score  code_challenge_multiple_choice_score  \
0                     NaN                                   NaN   
1                     NaN                                   NaN   
2                     NaN                                   NaN   
3                     NaN                                   NaN   
4                     NaN                                   NaN   

   code_challenge_final_score  code_challenge_cheat_tab_leaving_(no.)  \
0                         NaN                                     NaN   
1                         NaN                                     NaN   
2                         NaN                                     NaN   
3                         NaN                                     NaN   
4                         NaN                                     NaN   

  code_challenge_cheat_plagiarism code_challenge_cheat_pasted_code  \
0                             NaN                              NaN   
1                             NaN                       

In [24]:
not_applied = []

for row in applications_df["status"]:
    if row != "Applied":
        not_applied.append(row)

not_applied

[]

In [25]:
not_null = []

for row_number, row_item in enumerate(applications_df["english_proficiency"]):
    if row_item != "I can read, write and speak English easily":
        not_null.append({row_number: row_item})

not_null

[{20: 'I can read, write and speak English but it isnt easy'},
 {112: nan},
 {160: 'I can read, write and speak English but it isnt easy'},
 {193: 'I can read, write and speak English but it isnt easy'},
 {196: 'I can read, write and speak English but it isnt easy'},
 {200: 'I can read, write and speak English but it isnt easy'},
 {320: 'I struggle to read, write and speak english'},
 {323: 'I can read, write and speak English but it isnt easy'},
 {330: 'I can read, write and speak English but it isnt easy'},
 {341: 'I can read, write and speak English but it isnt easy'},
 {352: 'I can read, write and speak English but it isnt easy'},
 {444: 'I can read, write and speak English but it isnt easy'},
 {451: 'I can read, write and speak English but it isnt easy'},
 {497: 'I can read, write and speak English but its difficult'},
 {564: 'I can read, write and speak English but it isnt easy'},
 {574: 'I can read, write and speak English but it isnt easy'},
 {588: 'I can read, write and speak 

In [26]:
for column_name in non_mandatory:
    applications_df = applications_df.drop(columns=[column_name])

applications_df.head(5)

application_id  learner_id gender           race    nationality  \
0           41581         572    Man  African/Black  South African   
1           73815         857  Woman  African/Black  South African   
2           46124         908    Man  African/Black  South African   
3          254193         928    Man  African/Black  South African   
4           76156        1132  Woman  African/Black  South African   

  country_of_residence         registration_date          application_date  \
0         South Africa  2024-05-11T13:27:10.490Z  2025-03-22T12:03:12.669Z   
1         South Africa  2024-05-12T18:55:56.670Z  2025-07-21T07:32:00.689Z   
2         South Africa  2024-05-13T03:47:02.685Z  2025-04-15T22:45:48.117Z   
3         South Africa  2024-05-13T06:12:08.009Z  2025-09-19T13:47:28.466Z   
4         South Africa  2024-05-13T14:32:21.448Z  2025-08-03T08:55:31.590Z   

   aptitude_total_score  literacy_score  logic_score  sequence_score  \
0                    48               8            9               7   
1                    51               9            8               9   
2                    43               9            7               6   
3                    48               9            9               7   
4                    57               9           10               9   

   problem_solving_score  numeracy_score  statistics_score  \
0                      7               7                10   
1                      9               8                 8   
2                      3               9                 9   
3                      9               9                 5   
4                     10              10                 9   

  has_coding_experience                    programming_language_experience  \
0                 False                                             [null]   
1                  True           ["C++","HTML/CSS","JavaScript","Python"]   
2                 False                                             [null]   
3                  True  ["HTML/CSS","JavaScript","Other (please specif...   
4                  True                       ["HTML/CSS","Java","Python"]   

  other_programming_language_experince_specify learner_holds_coding_certs  \
0                                          NaN                        NaN   
1                                          NaN                      False   
2                                          NaN                        NaN   
3                                          NaN                       True   
4                                          NaN                       True   

              learner_coding_certs_list                    learner_program  \
0                                   NaN  Umuzi Programmes Application Form   
1                                   NaN  Umuzi Programmes Application Form   
2                                   NaN  Umuzi Programmes Application Form   
3  IITSA acredited certificate in React  Umuzi Programmes Application Form   
4           FNB APP ACADEMY CERTIFICATE  Umuzi Programmes Application Form   

                              learner_career_streams  \
0                     ["Full Stack Web Development"]   
1  ["Full Stack Web Development","Software Engine...   
2  ["Data Science ","Full Stack Web Development",...   
3  ["Full Stack Web Development","Project Managem...   
4                     ["Full Stack Web Development"]   

                      computer_access_type     employment_type  \
0                   I have my own computer  Seeking employment   
1  I have my own computer but it is faulty  Seeking employment   
2                   I have my own computer            Studying   
3                   I have my own computer  Seeking employment   
4                   I have my own computer            Studying   

                          english_proficiency highest_education_level  \
0  I can read, write and speak English easily             High School   
1  I can read, write and s

In [27]:
applications_df.shape

(599, 33)